File: hard76.cpp
//
// Created by hello on 8/14/22.
//

#include <iostream>
#include <unordered_map>

using namespace std;
string minWindow(string s, string t) {
    if(t.length()>s.length())return "";
    unordered_map<char,int>need,window;
    for(char c:t){need[c]++;}
    int left=0,right=0;
    int valid=0;
    int len=INT_MAX,start=0;   //最小子串
    while (right<s.length()){
        char c=s[right];
        right++;
        if(need.count(c)){
            window[c]++;
            if(window[c]==need[c]){
                valid++;
            }
        }
        while (valid==need.size()){   //左边进行收缩
            if(len>right-left){
                len=right-left;
                start=left;
            }
            char d=s[left];
            if(need.count(d)){
                if(need[d]==window[d]){
                    valid--;
                }
                window[d]--;
            }
            left++;
        }
    }
    return len==INT_MAX?"":s.substr(start,len);
}
int main(){
//    string ans=minWindow("ADOBECODEBANC","ABC");
    string ans=minWindow("a","aa");
    cout<<"ans:"<<ans<<endl;
}

File: hard76.java
package nopart;

import java.util.HashMap;

public class hard76 {
    public static void main(String[] args) {
//        String  ans=minWindow("ADOBECODEBANC","ABC");
        String  ans=minWindow("a","aa");
        System.out.println(ans);
    }

    public static String minWindow(String s, String t) {
        HashMap<Character,Integer>window=new HashMap<>();
        HashMap<Character,Integer>need=new HashMap<>();
        for(int i=0;i<t.length();i++){
            need.put(t.charAt(i),need.getOrDefault(t.charAt(i),0)+1);
        }
        int start=0;
        int len=Integer.MAX_VALUE;
        int left=0,right=0;
        int valid=0;
        while (right<s.length()){
            char c=s.charAt(right);
            right++;
            if(need.containsKey(c)){
                window.put(c,window.getOrDefault(c,0)+1);
                if(need.get(c).equals(window.get(c))){
                    valid++;
                }
            }
            while (valid==need.size()){
                if(right-left<len){
                    len=right-left;
                    System.out.println(len);
                    start=left;
                }
                char d=s.charAt(left);
                left++;
                if(need.containsKey(d)){
                    if(window.get(d).equals(need.get(d))){
                        valid--;             //神来之笔，放在window.get(c)-1之前
                    }
                    window.put(d,window.get(d)-1);
                }
            }
        }
        return len==Integer.MAX_VALUE?"":s.substring(start,start+len);
    }
}
























